src/snowflake/connector/cache.py

```diff
@@ -0,0 +1,264 @@
+#
+# Copyright (c) 2012-2021 Snowflake Computing Inc. All rights reserved.
+#
+
+from __future__ import annotations
+
+import datetime
+from collections.abc import Iterator
+from threading import Lock
+from typing import Generic, TypeVar
+
+from typing_extensions import NamedTuple, Self
+
+now = datetime.datetime.now
+
+T = TypeVar("T")
+
+
+class CacheEntry(NamedTuple, Generic[T]):
+    expiry: datetime.datetime
+    entry: T
+
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+def is_expired(d: datetime.datetime) -> bool:
+    return now() >= d
+
+
+class SFDictCache(Generic[K, V]):
```

Comment:
is the cache expected to be internally used only?

my concern is that improper usage pattern leads to memory monotonically increment (as we're only doing lazy deletion here) and after long running, users encounter out of memory error.

or this is not a valid concern?

Reply:
Internal only, yes. This exact idea was discussed in the design doc for the project, please take a look at that

Reply:
yes, I remember that. we know there's a fixed amount of cache and small size.